File: java-project-standard/assets/templates/adapters/src/test/java/__PKG_PATH__/adapters/ProviderConformanceTest.java
package __PKG__.adapters;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import __PKG__.domain.Embedder;
import __PKG__.domain.Llm;
import __PKG__.domain.ProviderException;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;
import org.junit.jupiter.api.Test;

/**
 * Conformance kit: every type claiming to implement a port (the mock and any real backend) runs the
 * same behavioral invariants -- hot-swapping is only safe when all plugs behave alike. Add a real
 * adapter's supplier to the maps below and it reuses these assertions unchanged.
 */
class ProviderConformanceTest {
  private static final Map<String, Supplier<Embedder>> EMBEDDERS =
      Map.of("MockEmbedder", MockEmbedder::new);
  private static final Map<String, Supplier<Llm>> LLMS = Map.of("MockLlm", MockLlm::new);

  @Test
  void embedderIsDeterministic() throws ProviderException {
    for (var entry : EMBEDDERS.entrySet()) {
      Embedder embedder = entry.getValue().get();
      List<double[]> first = embedder.embed(List.of("hello", "world"));
      List<double[]> second = embedder.embed(List.of("hello", "world"));
      assertEquals(first.size(), second.size(), entry.getKey());
      for (int i = 0; i < first.size(); i++) {
        assertArrayEquals(first.get(i), second.get(i), entry.getKey() + " must be deterministic");
      }
    }
  }

  @Test
  void embedderPreservesInputCount() throws ProviderException {
    for (var entry : EMBEDDERS.entrySet()) {
      assertEquals(3, entry.getValue().get().embed(List.of("a", "b", "c")).size(), entry.getKey());
    }
  }

  @Test
  void embedderReturnsNonEmptyVectors() throws ProviderException {
    for (var entry : EMBEDDERS.entrySet()) {
      for (double[] vector : entry.getValue().get().embed(List.of("x"))) {
        assertTrue(vector.length > 0, entry.getKey());
      }
    }
  }

  @Test
  void llmReturnsNonEmptyCompletion() throws ProviderException {
    for (var entry : LLMS.entrySet()) {
      assertTrue(!entry.getValue().get().complete("ping").isEmpty(), entry.getKey());
    }
  }
}

File: android-project-standard/assets/templates/adapters/src/test/kotlin/__PKG_PATH__/adapters/ProviderConformanceTest.kt
// 一致性契约(conformance kit):任何号称实现了某 port 的类型(Mock 与真实后端)都必须
// 跑过同一组行为不变量 —— 可插拔只在所有插头行为一致时才安全。
//
// 真实 adapter(启用 gated module / 有 key 时)把工厂加进对应 list 即复用同一组断言。

package __APP_ID__.adapters

import __APP_ID__.domain.Embedder
import __APP_ID__.domain.Llm
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ProviderConformanceTest {
    private val embedders: List<Pair<String, () -> Embedder>> =
        listOf(
            "MockEmbedder" to { MockEmbedder() },
            // "OpenAiEmbedder" to { OpenAiEmbedder() },
        )
    private val llms: List<Pair<String, () -> Llm>> =
        listOf(
            "MockLlm" to { MockLlm() },
        )

    @Test
    fun embedderIsDeterministic() =
        runTest {
            for ((name, make) in embedders) {
                val embedder = make()
                val first = embedder.embed(listOf("hello", "world"))
                val second = embedder.embed(listOf("hello", "world"))
                assertEquals(first, second, "$name 应确定性:同输入同输出")
            }
        }

    @Test
    fun embedderPreservesInputCount() =
        runTest {
            for ((name, make) in embedders) {
                val vectors = make().embed(listOf("a", "b", "c"))
                assertEquals(3, vectors.size, "$name 应保持输入数量")
            }
        }

    @Test
    fun embedderReturnsNonEmptyVectors() =
        runTest {
            for ((name, make) in embedders) {
                val vectors = make().embed(listOf("x"))
                assertTrue(vectors.all { it.isNotEmpty() }, "$name 不应返回空向量")
            }
        }

    @Test
    fun llmReturnsNonEmptyCompletion() =
        runTest {
            for ((name, make) in llms) {
                assertTrue(make().complete("ping").isNotEmpty(), "$name 应返回非空回答")
            }
        }
}
